### PR TITLE
Modernize ResetPassword ViewModel in Kotlin

### PR DESCRIPTION
**Pull request by miga9**
This is a PR for modernizing ResetPasswordViewModel **in Kotlin** according to your latest VM architecture doc.

### CODE REVIEW

File: app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+      return this.resetSuccess
+    }
+
+    override fun resetError(): Observable<String> {
+      return this.resetError
+        .takeUntil(this.resetSuccess)
+        .map { it.errorMessage() }
+    }
+  }
+}
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt]: newline would be nice here!

File: app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    /** Emits a boolean that determines if the form validation is passing. */
+    fun isFormValid(): Observable<Boolean>
+
+    /** Emits when password reset is completed successfully. */
+    fun resetSuccess(): Observable<Void>
+  }
+
+  interface Errors {
+    /** Emits when password reset fails. */
+    fun resetError(): Observable<String>
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt]: same feedback here, let's consolidate Errors to Outputs

**Comment by miga9** [comment]
@luoser Thanks for your feedback! ;)

File: app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    init {
+      this.email
+        .map { StringUtils.isEmail(it) }
+        .compose(bindToLifecycle())
+        .subscribe(this.isFormValid)
+
+      this.email
+        .compose<String>(Transformers.takeWhen(this.resetPasswordClick))
+        .switchMap { this.submitEmail(it) }
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt]: same here!
```
        .switchMap(this::submitEmail)
```

File: app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    private val isFormValid = PublishSubject.create<Boolean>()
+    private val resetSuccess = PublishSubject.create<Void>()
+    private val resetError = PublishSubject.create<ErrorEnvelope>()
+
+    val inputs: Inputs = this
+    val outputs: Outputs = this
+
+    init {
+      this.email
+        .map { StringUtils.isEmail(it) }
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt]: one cool FP trick you can do here is to make this function call point free, e.g.
```
        .map(StringUtils::isEmail)
```

Comment (miga9) [reply app/src/main/java/com/kickstarter/viewmodels/ResetPasswordViewModel.kt]: Thanks for pointing this out! I read about pointful vs point-free style.

**Pull request merged**